### ISSUE TITLE
node change, split create permanent links between copy of msg attributes during the flow process

### ISSUE THREAD

**Issue by abruno06** [node-red/node-red#3144]
### Current Behavior

Hello 

I'm experimenting a very unexpected situation in my production flow, that I have emulated on the attach flow using Nodered node only to have the exact same result.

While using split and change nodes, I observe a kind of link between fields in the msg processing.
msg is inject with 
![image](https://user-images.githubusercontent.com/17586751/133579577-426439be-aee3-4d94-bb48-c0a9db82859c.png)
as msg.variables
then 
![image](https://user-images.githubusercontent.com/17586751/133579720-105f6df5-05f8-4fd9-bf6b-dcf3d801ce73.png)
msg.payload is create from msg.variables.input

Then the payload (array) is split in two msg using the split node.

the next change node create the msg.v element
![image](https://user-images.githubusercontent.com/17586751/133580015-c6c5b8ca-d033-43ab-8d35-1d5f43d45dad.png)

I then add another field in the payload

![image](https://user-images.githubusercontent.com/17586751/133580145-ce913a57-5b9a-440a-9bae-e1912afb0688.png)

and here the results

![image](https://user-images.githubusercontent.com/17586751/133580300-2595b398-458e-4bac-b9b6-ca3fb7cc6148.png)

as you can see, only by adding something on the payload, the msg.v get the change, but also the msg.variables.input[0] (which is the element split inject) get the added element.
similar happen on the second msg from the slit

I have attempt a workaround
![image](https://user-images.githubusercontent.com/17586751/133580836-46ede859-970a-4c6d-a5a7-9dacb009e239.png)
by deleting the payload and recreating it from msg.v

then adding another element 
![image](https://user-images.githubusercontent.com/17586751/133581175-453cd894-31e8-40a1-8840-c7e76acf4d1f.png)

but same results
![image](https://user-images.githubusercontent.com/17586751/133581125-5a870bcd-e904-4e77-b908-389a975ad369.png)
see how element1 get propagated to v and the variables.input 
I also test deletion 
![image](https://user-images.githubusercontent.com/17586751/133581336-050d6198-7663-49b5-956e-76d7cc0af596.png)

and same cascaded effect
![image](https://user-images.githubusercontent.com/17586751/133581481-77aac5b2-df71-42b7-88c6-fe085ea42e28.png)

This is the very first time I'm experimenting this situation, not sure if this is an expected behavior, but this have serious impact on flow processing.
I also test using jsonata
![image](https://user-images.githubusercontent.com/17586751/133582074-f99edf54-3835-4d61-b278-367bdc8986df.png)
but same behavior

### Expected Behavior

msg is not expected to have links between msg attribute copy and especially in between nodes

### Steps To Reproduce

execute the flow below and look at the debug message

### Example flow

```
[
    {
        "id": "2ecc64a067b132d9",
        "type": "inject",
        "z": "12b77e636fcff8f5",
        "name": "",
        "props": [
            {
                "p": "variables",
                "v": "{\"input\":[{\"key1\":\"value1\"},{\"key1\":\"value2\"}]}",
                "vt": "json"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "x": 180,
        "y": 200,
        "wires": [
            [
                "7a4292b4f8cd0680"
            ]
        ]
    },
    {
        "id": "7a4292b4f8cd0680",
        "type": "change",
        "z": "12b77e636fcff8f5",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "variables.input",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 360,
        "y": 200,
        "wires": [
            [
                "7531cd61d44ddac6"
            ]
        ]
    },
    {
        "id": "7531cd61d44ddac6",
        "type": "split",
        "z": "12b77e636fcff8f5",
        "name": "",
        "splt": "\\n",
        "spltType": "str",
        "arraySplt": 1,
        "arraySpltType": "len",
        "stream": false,
        "addname": "",
        "x": 530,
        "y": 200,
        "wires": [
            [
                "d8ea82228d15c954"
            ]
        ]
    },
    {
        "id": "d8ea82228d15c954",
        "type": "change",
        "z": "12b77e636fcff8f5",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "v",
                "pt": "msg",
                "to": "payload",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 680,
        "y": 200,
        "wires": [
            [
                "4e82a5a95a9e702b"
            ]
        ]
    },
    {
        "id": "4cf42df11dcda1cd",
        "type": "debug",
        "z": "12b77e636fcff8f5",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 890,
        "y": 140,
        "wires": []
    },
    {
        "id": "4e82a5a95a9e702b",
        "type": "change",
        "z": "12b77e636fcff8f5",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload.add",
                "pt": "msg",
                "to": "hello",
                "tot": "str"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 860,
        "y": 200,
        "wires": [
            [
                "4cf42df11dcda1cd",
                "2142cd19933eec6d"
            ]
        ]
    },
    {
        "id": "6618cc1a3da6455c",
        "type": "debug",
        "z": "12b77e636fcff8f5",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 910,
        "y": 300,
        "wires": []
    },
    {
        "id": "2142cd19933eec6d",
        "type": "change",
        "z": "12b77e636fcff8f5",
        "name": "",
        "rules": [
            {
                "t": "delete",
                "p": "payload",
                "pt": "msg"
            },
            {
                "t": "set",
                "p": "payload.vcopy",
                "pt": "msg",
                "to": "v",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 400,
        "y": 300,
        "wires": [
            [
                "9b9668bc38d9d2ac"
            ]
        ]
    },
    {
        "id": "9b9668bc38d9d2ac",
        "type": "change",
        "z": "12b77e636fcff8f5",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload.vcopy.element1",
                "pt": "msg",
                "to": "AddMore",
                "tot": "str"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 680,
        "y": 300,
        "wires": [
            [
                "6618cc1a3da6455c",
                "45283a9c974e711e"
            ]
        ]
    },
    {
        "id": "45283a9c974e711e",
        "type": "change",
        "z": "12b77e636fcff8f5",
        "name": "",
        "rules": [
            {
                "t": "delete",
                "p": "payload.vcopy.key1",
                "pt": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 450,
        "y": 380,
        "wires": [
            [
                "f04ee88eb2d20f5a"
            ]
        ]
    },
    {
        "id": "f04ee88eb2d20f5a",
        "type": "debug",
        "z": "12b77e636fcff8f5",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 910,
        "y": 380,
        "wires": []
    }
]
```

### Environment

- Node-RED version: 2.0.6
- Node.js version: v14.17.6
- npm version: v6.14.15
- Platform/OS: CentOS
- Browser: Firefox

**Comment by thechane** [comment]
There seems to be something up with the msg cloning. This is an ugly workaround that prevents msg.variables from being updated with those after split node keys by cloning manually. 

`
[
    {
        "id": "2ecc64a067b132d9",
        "type": "inject",
        "z": "6dc0dab2.712454",
        "name": "",
        "props": [
            {
                "p": "variables",
                "v": "{\"input\":[{\"key1\":\"value1\"},{\"key1\":\"value2\"}]}",
                "vt": "json"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payloadType": "str",
        "x": 150,
        "y": 280,
        "wires": [
            [
                "413ba920b9d666ed"
            ]
        ]
    },
    {
        "id": "7a4292b4f8cd0680",
        "type": "change",
        "z": "6dc0dab2.712454",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "variables.input",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 350,
        "y": 300,
        "wires": [
            [
                "7531cd61d44ddac6"
            ]
        ]
    },
    {
        "id": "7531cd61d44ddac6",
        "type": "split",
        "z": "6dc0dab2.712454",
        "name": "",
        "splt": "\\n",
        "spltType": "str",
        "arraySplt": 1,
        "arraySpltType": "len",
        "stream": false,
        "addname": "",
        "x": 540,
        "y": 280,
        "wires": [
            [
                "a03a734d52fbbfbe"
            ]
        ]
    },
    {
        "id": "d8ea82228d15c954",
        "type": "change",
        "z": "6dc0dab2.712454",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "v",
                "pt": "msg",
                "to": "payload",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 680,
        "y": 300,
        "wires": [
            [
                "4e82a5a95a9e702b"
            ]
        ]
    },
    {
        "id": "4cf42df11dcda1cd",
        "type": "debug",
        "z": "6dc0dab2.712454",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 900,
        "y": 220,
        "wires": []
    },
    {
        "id": "4e82a5a95a9e702b",
        "type": "change",
        "z": "6dc0dab2.712454",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload.add",
                "pt": "msg",
                "to": "hello",
                "tot": "str"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 870,
        "y": 280,
        "wires": [
            [
                "4cf42df11dcda1cd",
                "2142cd19933eec6d"
            ]
        ]
    },
    {
        "id": "6618cc1a3da6455c",
        "type": "debug",
        "z": "6dc0dab2.712454",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 920,
        "y": 380,
        "wires": []
    },
    {
        "id": "2142cd19933eec6d",
        "type": "change",
        "z": "6dc0dab2.712454",
        "name": "",
        "rules": [
            {
                "t": "delete",
                "p": "payload",
                "pt": "msg"
            },
            {
                "t": "set",
                "p": "payload.vcopy",
                "pt": "msg",
                "to": "v",
                "tot": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 410,
        "y": 380,
        "wires": [
            [
                "9b9668bc38d9d2ac"
            ]
        ]
    },
    {
        "id": "9b9668bc38d9d2ac",
        "type": "change",
        "z": "6dc0dab2.712454",
        "name": "",
        "rules": [
            {
                "t": "set",
                "p": "payload.vcopy.element1",
                "pt": "msg",
                "to": "AddMore",
                "tot": "str"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 690,
        "y": 380,
        "wires": [
            [
                "6618cc1a3da6455c",
                "45283a9c974e711e"
            ]
        ]
    },
    {
        "id": "45283a9c974e711e",
        "type": "change",
        "z": "6dc0dab2.712454",
        "name": "",
        "rules": [
            {
                "t": "delete",
                "p": "payload.vcopy.key1",
                "pt": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 460,
        "y": 460,
        "wires": [
            [
                "f04ee88eb2d20f5a"
            ]
        ]
    },
    {
        "id": "f04ee88eb2d20f5a",
        "type": "debug",
        "z": "6dc0dab2.712454",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 920,
        "y": 460,
        "wires": []
    },
    {
        "id": "413ba920b9d666ed",
        "type": "function",
        "z": "6dc0dab2.712454",
        "name": "",
        "func": "msg.payload = JSON.parse(JSON.stringify(msg.variables.input));\nreturn msg;",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 360,
        "y": 220,
        "wires": [
            [
                "7531cd61d44ddac6"
            ]
        ]
    },
    {
        "id": "a03a734d52fbbfbe",
        "type": "function",
        "z": "6dc0dab2.712454",
        "name": "",
        "func": "msg.v = JSON.parse(JSON.stringify(msg.payload));\nreturn msg;",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 660,
        "y": 220,
        "wires": [
            [
                "4e82a5a95a9e702b"
            ]
        ]
    }
]
`

**Comment by abruno06** [comment]
I create a small node (node-red-contrib-breaker) that integrate the workaround from @thechane. I did couple of extra trial and found there is definitively something around the RED.node.cloneMessage() that will need more investigation; this is probably around the clone function itself (I have try to use clone, but by itself, without workaournd, it suffer same issue than clonedeep; you can produce link between msg fields). definitively something that need more deep dive.

**Comment by knolleary** [comment]
I think this is 'expected' behaviour. Although it's taken a while to pick through all of the steps you outline. With multiple Debug nodes in the flow, its hard to know what I'm meant to be looking for in its output, even with your screenshots.


At the start you have:
```
{
   variables: { "input":  [ { "k1", "v1" }, { "k2": "v2" } ] }
}
```

You then set `msg.payload` to `msg.variables.input`:
```
{
   payload: [ { "k1", "v1" }, { "k2": "v2" } ] ,
   variables: { "input":  [ { "k1", "v1" }, { "k2": "v2" } ] }
}
```

But, due to the way JavaScript copies references not objects, `msg.payload` and `msg.variables.input` point to the *same* object in memory.

You then pass it through the Split node, so you get one message for each `msg.payload` entry:

```
{
   payload: {"k1", "v1"},
   variables: {"input":  [ { "k1", "v1" }, { "k2": "v2" } ] }
}
```

Again, `msg.payload` and `msg.variables.input[0]` are pointers to the same object, so any modification to one will modify the other.

I don't see anything that is specific to the Split node here - it is doing the right thing. The issue is due to the reference between  `msg.payload` to `msg.variables.input` in the message passed to Split.

The workaround is to replace the Change node at the start of the flow with a Function node that does:
```
msg.payload = RED.util.cloneMessage(msg.variables.input)
return msg;
```

This will create a complete copy of the property.

This is true whenever you use the Change node to move/copy a property that is an Object or Array type.


We need to find a way to make this copy-by-reference behaviour more apparent in the Change node. We could consider adding an option to the Change node to clone rather than just set the property - the hard part is finding a way to expose it that remain accessible to users who don't need to know/care about copy-by-reference vs copy-by-value.

**Comment by natcl** [comment]
I have face similar issues in the past and it's not always easy to debug. I agree having an option in the change node would be ideal !

**Comment by abruno06** [comment]
Hello
I'm agree about the Change node; there is no cloneMessage in the node which could lead to some side effect due to NodeJS behavior.
but I have issue with the Split node also and this node have the cloneMessage. the Array is split by split node which then should not be linked anymore with the original msg element but this happened and this is what I have hard time to explain.

This is one of the reason I have created the node-red-contrib-breaker node to trace this with more granularity.
result of my test are 
- cloneMessage, still have link in the object
- use of lodash.clonedeep, still link in the object 
- use JSON.stringify : OK but have major drawback.
I come to conclusion that the issue was related to clonedeep not working as expected but again after spending time on the code on it do not see something that could lead to this situation are object are clearly recreated and reference broken.

I still have one option to explore is the asynchronous code execution that could lead to some leakage. 
I observer this issue on others code I did that also include a clone function and observe strange behavior while working in asynchronous or promise functions.

Outside of this, having the option to make deep clone in change will be nice

**Comment by knolleary** [comment]
It would be helpful to have a clear description of the split node issue once the fact the change node doesn't clone has been removed from the equation.

The update to the change node, to add the clone option, will still use the cloneMessage utility function to do the cloning. So if you think there is an issue with that function, it would be helpful to know.

If an object has properties that reference each other, cloning the message will break the link between the original message and the clone, but cross references within the message will still exist.

**Comment by knolleary** [comment]
With the new deep copy option available in 2.1, I'm going to close this issue.
